crypto/asn1_dsa.c

```diff
@@ -0,0 +1,298 @@
+        }
+    }
+    *pcont_len = n;
+    *ppin = in;
+    return consumed;
+}
+
+/*
+ * Decodes a single ASN.1 INTEGER value from *ppin, which must be DER encoded,
+ * updates n with the decoded value, and, if successful, increments *ppin past
+ * the data that was consumed. The BIGNUM, n, must have already been allocated
+ * by calling BN_new().
+ * An attempt to consume more than len bytes results in an error.
+ * Returns the number of bytes of input consumed or 0 if an error occurs.
+ *
+ * If the buffer is supposed to only contain a single INTEGER value with no
+ * trailing garbage then it is up to the caller to verify that all bytes
+ * were consumed.
+ */
+size_t decode_der_integer(BIGNUM *n, const unsigned char **ppin, size_t len)
+{
+    const unsigned char *in = *ppin;
+    size_t consumed;
+    size_t c;
+    size_t cont_len;
+
+    if (len < 1 || n == NULL || *in++ != ID_INTEGER)
+        return 0;
+    consumed = 1;
+    if ((c = decode_der_length(&cont_len, &in, len - consumed)) == 0)
+        return 0;
+    consumed += c;
+    /* Check for a positive INTEGER with valid content encoding and decode. */
+    if (cont_len > len - consumed || cont_len < 1 || (in[0] & 0x80) != 0
+            || (cont_len >= 2 && in[0] == 0 && (in[1] & 0x80) == 0)
+            || BN_bin2bn(in, (int)cont_len, n) == NULL)
+        return 0;
+    in += cont_len;
+    consumed += cont_len;
+    *ppin = in;
+    return consumed;
+}
+
+static size_t decode_dsa_sig_content(BIGNUM *r, BIGNUM *s,
+                                     const unsigned char **ppin, size_t len)
+{
+    const unsigned char *in = *ppin;
+    size_t consumed = 0;
+    size_t c;
+
+    if ((c = decode_der_integer(r, &in, len - consumed)) == 0)
+        return 0;
+    consumed += c;
+    if ((c = decode_der_integer(s, &in, len - consumed)) == 0)
+        return 0;
+    consumed += c;
+    *ppin = in;
+    return consumed;
+}
+
+/*
+ * Decodes a single DSA-Sig-Value or ECDSA-Sig-Value from *ppin, which must be
+ * DER encoded, updates r and s with the decoded values, and increments *ppin
+ * past the data that was consumed. The BIGNUMs, r and s, must have already
+ * been allocated by calls to BN_new().
+ * An attempt to consume more than len bytes results in an error.
+ * Returns the number of bytes of input consumed or 0 if an error occurs.
+ *
+ * If the buffer is supposed to only contain a single [EC]DSA-Sig-Value with no
+ * trailing garbage then it is up to the caller to verify that all bytes
+ * were consumed.
+ */
+size_t decode_der_dsa_sig(BIGNUM *r, BIGNUM *s, const unsigned char **ppin,
+                          size_t len)
+{
+    const unsigned char *in = *ppin;
+    size_t consumed;
+    size_t c;
+    size_t cont_len;
+
+    if (len < 1 || *in++ != ID_SEQUENCE)
+        return 0;
+    consumed = 1;
+    if ((c = decode_der_length(&cont_len, &in, len - consumed)) == 0)
+        return 0;
+    consumed += c;
+    if (cont_len > len - consumed
```

Comment:
same here - not sure size_t comparisons with len - consumed possibly being negative is a good idea.

Reply:
As for the encoders, in every decode function, and at each step, `consumed` is never greater than `len` so `len - consumed` can never be negative.
The value returned by a decode function is either 0 (for an error) or is less than or equal to it's len parameter. If the caller does `consumed += c;` then it is still true that `consumed` is less than or equal to `len`.